WordPress/gutenberg
# Edit Post: Add block management modal

Closes #14139

This pull request seeks to add a new "Manage Blocks" option to the editor menu which, when clicked, will present the user with an options modal allowing them to enable and disable blocks from being made available for use in the editor.

![Block Manager](https://user-images.githubusercontent.com/1779930/53759449-5f7fcd00-3e8e-11e9-8023-64b7ef9a62c8.png)

**Open Questions:**

_Usability Questions:_

- How should we handle block types which are either contextually available in the inserter, or are never made available? For example, child block types like "Column", deprecated block types like "Subheading", or inline block types like "Inline Image".
- If a theme or plugin [filters allowed block types](https://wordpress.org/gutenberg/handbook/designers-developers/developers/filters/block-filters/#hiding-blocks-from-the-inserter), how should this interoperate with the disabled block types? Should those excluded blocks be shown and made "perma-disabled" (i.e. the user can see the block type as disabled, but cannot enable it), should they not be present at all, or should they serve as a default where the user could re-enable them?
   - _Opinion:_ I don't think it would be advisable to allow a user to re-enable them, as plugins may make assumptions around block content structure around e.g. a custom post type.
- I may anticipate some feedback around whether the block disabled appearance in the manager is too subtle; if it needs to be changed, should we apply those changes consistently to the Inserter menu as well for disabled inserter blocks?
   - Similarly, there is a departure from design mockups in #14139 in the placement of the "Disable all" checkbox, where my current implementation chose to inherit UI from the Inserter/Transform for consistency's sake. Or would we prefer to break from those expectations for the "Disable all" button placement?
- How sticky should the setting be, and for whom should it apply? It is currently implemented as a user preference which persists across _post editor_ sessions. Notably, this means the disabled blocks would not be remembered for a future implementation of a block editor embedded in the widgets screen.
   - Responding to @mapk 's remark at https://github.com/WordPress/gutenberg/issues/14139#issuecomment-468716432 "I do love that if a block is in use, when clicking 'Disable all,' that particular block doesn't disable." : A problem is that this behavior is post-specific, which would conflict with the idea of the disabled block setting persisting across post editing sessions. Phrased differently, if I'm seeking to disable a particular block type, should I be conscious of the current post or not? It feels separate from the editing experience of a specific post.
   - Is it correct to think about this as a user preference, or would it be more appropriate as a site option? For example, a site administrator could foreseeably want to disable a particular block type for all users on a site. This could be considered as an additional enhancement, and if so, should it be deferred for a future implementation?

_Technical Questions:_

- Relating to the last top-level usability question above, this is implemented as a combination of preferences and components within `edit-post`. Depending how broadly this should apply, I think it could also be appropriate instead in `editor` or `block-editor`, or perhaps even its own separate module.
- While there's a good amount of reuse of existing components made possible, there's also some repetition and some questionable interfaces:
   - [`BlockTypesList`](https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/block-types-list/index.js):
      - Seems like a useful component, but I don't like how it's tied to the "inserter item" object shape, especially given the name of the component
         - Should the component be refactored to accept `blockTypes` as a prop? The inserter-unique behaviors are showing the "stacked" appearance for block types which can have children inserted within, and being able to make an item appear/act as disabled. Perhaps a pattern of function children could be leveraged to customize the behavior for each context?
      - Search behavior is duplicated from `InserterMenu` (normalizing search terms)
         - Should search implementation be embedded in the `BlockTypesList` component?
         - Should search be exposed as a selector(s) within e.g. `core/blocks` ?
      - While "disabled" could be a shared characteristic, the behavior is not the same. In the Inserter / Transforms menu, it disallows selection. In the Block Manager, it should act more as an active/inactive toggle


**Testing Instructions:**

Verify that you can disable/enable individual blocks/whole categories of blocks from the Block Manager, found under More Options > Manage Blocks.

Disabling a block should prevent it from being inserted within the post. (Note: It does not currently have an impact on blocks already within a post, even if those blocks become disabled by the user)
